pkg/appgw/backendaddresspools.go

```diff
@@ -7,82 +7,121 @@ package appgw
+		builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
+		glog.Warning(logLine)
+		return defaultPool
+	}
 
-				builder.backendPoolMap[backendID] = &addressPool
-				break
+	for _, subset := range endpoints.Subsets {
+		if _, portExists := getPorts(subset)[serviceBackendPair.BackendPort]; portExists {
+			poolName := generateAddressPoolName(backendID.serviceFullName(), backendID.Backend.ServicePort.String(), serviceBackendPair.BackendPort)
+			// The same service might be referenced in multiple ingress resources, this might result in multiple `serviceBackendPairMap` having the same service key but different
+			// ingress resource. Thus, while generating the backend address pool, we should make sure that we are generating unique backend address pools.
+			pool, ok := (*addressPools)[poolName]
+			if !ok {
+				pool = newPool(poolName, subset)
+				(*addressPools)[poolName] = pool
 			}
+			return pool
 		}
 	}
+	return nil
+}
+
+func (builder *appGwConfigBuilder) getServiceBackendPairMap() map[backendIdentifier]serviceBackendPortPair {
+	// TODO(draychev): deprecate the use of builder.serviceBackendPairMap
+	// Create this struct here instead of backendhttpsettings.go
+	return builder.serviceBackendPairMap
+}
 
-	backendPools := make([](network.ApplicationGatewayBackendAddressPool), 0)
-	for _, addressPool := range addressPools {
-		backendPools = append(backendPools, addressPool)
+func newPool(poolName string, subset v1.EndpointSubset) *n.ApplicationGatewayBackendAddressPool {
+	return &n.ApplicationGatewayBackendAddressPool{
+		Etag: to.StringPtr("*"),
+		Name: &poolName,
+		ApplicationGatewayBackendAddressPoolPropertiesFormat: &n.ApplicationGatewayBackendAddressPoolPropertiesFormat{
+			BackendAddresses: getAddressesForSubset(subset),
+		},
 	}
+}
 
-	builder.appGwConfig.BackendAddressPools = &backendPools
+func getAddressesForSubset(subset v1.EndpointSubset) *[]n.ApplicationGatewayBackendAddress {
+	// We make separate maps for IP and FQDN to ensure uniqueness within the 2 groups
+	// We cannot use ApplicationGatewayBackendAddress as it contains pointer to strings and the same IP string
+	// at a different address would be 2 unique keys.
+	addrSet := make(map[n.ApplicationGatewayBackendAddress]interface{})
+	ips := make(map[string]interface{})
+	fqdns := make(map[string]interface{})
+	for _, address := range subset.Addresses {
+		// prefer IP address
+		if len(address.IP) != 0 {
+			// address specified by ip
+			ips[address.IP] = nil
+		} else if len(address.Hostname) != 0 {
+			// address specified by hostname
+			fqdns[address.Hostname] = nil
+		}
+	}
 
-	return builder, nil
+	for ip := range ips {
+		addrSet[n.ApplicationGatewayBackendAddress{IPAddress: to.StringPtr(ip)}] = nil
+	}
+	for fqdn := range fqdns {
+		addrSet[n.ApplicationGatewayBackendAddress{IPAddress: to.StringPtr(fqdn)}] = nil
+	}
+	return getBackendAddressMapKeys(&addrSet)
```

Comment:
can we move the `getBackendAddressMapKeys` below this function ?
I like a flow where a function is used first to establish some context and then defined in the next section.